File: tests/oracle/printlanguage_atom_metadata_1204.cc
/*
 * PRINTLANGUAGE-ATOM-METADATA-0001
 *
 * Locked Ghidra 12.0.4 projection for PrintLanguage::pushAtom/emitAtom.
 * A hidden RPN operator forces the variable Atom through the RPN stack while
 * a recording emitter observes the exact tagVariable arguments.  Native
 * Varnode/PcodeOp pointers are normalized only to fixture creation-order IDs.
 */

#include <bits/stdc++.h>

// The fixture connects the Varnode to its real consuming input slot.  Ghidra
// normally performs this through Funcdata/PcodeOpBank; test-only private access
// keeps this language-independent fixture free of Architecture bootstrap.
#define private public
#include "printc.hh"
#include "op.hh"
#include "varnode.hh"
#undef private

namespace {

using namespace ghidra;

std::string hexText(const std::string &text) {
  static const char digits[] = "0123456789abcdef";
  std::string result;
  result.reserve(text.size() * 2);
  for (std::string::const_iterator iter = text.begin(); iter != text.end(); ++iter) {
    const unsigned char value = static_cast<unsigned char>(*iter);
    result.push_back(digits[value >> 4]);
    result.push_back(digits[value & 0xf]);
  }
  return result;
}

std::string colorName(EmitMarkup::syntax_highlight highlight) {
  if (highlight == EmitMarkup::const_color)
    return "const";
  if (highlight == EmitMarkup::no_color)
    return "none";
  throw LowlevelError("unexpected fixture highlight");
}

class RecordingEmit final : public EmitNoMarkup {
  const std::vector<const Varnode *> &varnodes;
  const std::vector<const PcodeOp *> &ops;
  std::vector<std::string> &events;
  int4 nextGroup;

  std::string varnodeId(const Varnode *vn) const {
    if (vn == (const Varnode *)0)
      return "none";
    for (size_t i = 0; i < varnodes.size(); ++i) {
      if (varnodes[i] == vn)
        return "obj" + std::to_string(i);
    }
    throw LowlevelError("unregistered fixture Varnode pointer");
  }

  std::string opId(const PcodeOp *op) const {
    if (op == (const PcodeOp *)0)
      return "none";
    for (size_t i = 0; i < ops.size(); ++i) {
      if (ops[i] == op)
        return "obj" + std::to_string(varnodes.size() + i);
    }
    throw LowlevelError("unregistered fixture PcodeOp pointer");
  }

  void append(const std::string &kind, const std::string &text,
              EmitMarkup::syntax_highlight highlight,
              const Varnode *vn, const PcodeOp *op,
              const std::string &group) {
    std::ostringstream line;
    line << "event=" << events.size()
         << "|kind=" << kind
         << "|text_hex=" << hexText(text)
         << "|highlight=" << colorName(highlight)
         << "|vn=" << varnodeId(vn)
         << "|op=" << opId(op)
         << "|group=" << group;
    events.push_back(line.str());
  }

public:
  RecordingEmit(const std::vector<const Varnode *> &vnOrder,
                const std::vector<const PcodeOp *> &opOrder,
                std::vector<std::string> &output)
      : varnodes(vnOrder), ops(opOrder), events(output), nextGroup(0) {}

  int4 openGroup(void) override {
    const int4 id = nextGroup++;
    append("open_group", "", no_color, (const Varnode *)0,
           (const PcodeOp *)0, "g" + std::to_string(id));
    return id;
  }

  void closeGroup(int4 id) override {
    if (id < 0 || id >= nextGroup)
      throw LowlevelError("unknown fixture group id");
    append("close_group", "", no_color, (const Varnode *)0,
           (const PcodeOp *)0, "g" + std::to_string(id));
  }

  void tagVariable(const std::string &name, syntax_highlight highlight,
                   const Varnode *vn, const PcodeOp *op) override {
    append("variable", name, highlight, vn, op, "none");
  }

  void print(const std::string &data,
             syntax_highlight highlight = no_color) override {
    append("syntax", data, highlight, (const Varnode *)0,
           (const PcodeOp *)0, "none");
  }
};

class FixturePrintC final : public PrintC {
public:
  explicit FixturePrintC(RecordingEmit *recorder)
      : PrintC((Architecture *)0, "printlanguage-atom-metadata-1204") {
    delete emit;
    emit = recorder;
  }

  void emitFixture(const Varnode *vn, const PcodeOp *consumer) {
    const std::string constantText("'\\0'");
    const std::string syntaxText(";");
    const OpToken fixtureHidden = {
        "", "", 1, 70, false, OpToken::hiddenfunction, 0, 0,
        (OpToken *)0};

    // The locked PrintC hidden token is a stage-1 operator.  An explicit copy
    // makes the RPN input manifest self-contained while exercising the
    // non-empty stack branch without adding an unrelated visible event.
    pushOp(&fixtureHidden, consumer);
    pushAtom(Atom(constantText, vartoken, EmitMarkup::const_color,
                  consumer, vn, 0));

    // Repeating the same annotations proves identity is stable across events,
    // rather than a per-event pointer rendering artifact.
    pushOp(&fixtureHidden, consumer);
    pushAtom(Atom(constantText, vartoken, EmitMarkup::const_color,
                  consumer, vn, 0));

    // The no-annotation Atom constructor deliberately leaves the union/op
    // members unspecified.  emitAtom's syntax branch must not read them.
    pushAtom(Atom(syntaxText, syntax, EmitMarkup::no_color));

    if (!revpol.empty() || !nodepend.empty() || pending != 0)
      throw LowlevelError("Ghidra fixture did not drain RPN state");
  }
};

} // namespace

int main() {
  Varnode constant(1, Address((AddrSpace *)0, 0), (Datatype *)0);
  constant.flags |= Varnode::constant;
  constant.nzm = 0;
  PcodeOp consumer(1, SeqNum(Address((AddrSpace *)0, 0), 43));
  consumer.setInput(&constant, 0);
  if (!constant.isConstant() || constant.getOffset() != 0 ||
      consumer.getIn(0) != &constant)
    throw LowlevelError("fixture consuming-op precondition failed");

  const std::vector<const Varnode *> varnodeOrder(1, &constant);
  const std::vector<const PcodeOp *> opOrder(1, &consumer);
  std::vector<std::string> events;
  {
    RecordingEmit *recorder = new RecordingEmit(varnodeOrder, opOrder, events);
    FixturePrintC printer(recorder);
    printer.emitFixture(&constant, &consumer);
  }

  std::cout
      << "schema=1|fixture=PRINTLANGUAGE-ATOM-METADATA-0001"
      << "|oracle=e40ed13014025f82488b1f8f7bca566894ac376b"
      << "|covered_projection=MATCH|overall=MISMATCH\n";
  for (std::vector<std::string>::const_iterator iter = events.begin();
       iter != events.end(); ++iter)
    std::cout << *iter << '\n';
  return 0;
}

File: tests/oracle/printlanguage_atom_metadata_1204.rs
//! PRINTLANGUAGE-ATOM-METADATA-0001 Rugra comparand.
//!
//! The IDs below are fixture creation-order identities corresponding to the
//! native-pointer normalization performed by the locked C++ oracle.

use rugra::prettyprint::Emit;
use rugra::printlanguage::{
    rpn_push_atom, rpn_push_op, Atom, NodePending, OpToken, ReversePolish, SyntaxHighlight,
    TagType, TokenType,
};

struct RecordingEmit {
    events: Vec<String>,
    next_group: i32,
}

impl Default for RecordingEmit {
    fn default() -> Self {
        Self {
            events: Vec::new(),
            next_group: 0,
        }
    }
}

fn hex_text(text: &str) -> String {
    text.as_bytes()
        .iter()
        .map(|byte| format!("{byte:02x}"))
        .collect()
}

fn color_name(highlight: SyntaxHighlight) -> &'static str {
    match highlight {
        SyntaxHighlight::ConstColor => "const",
        SyntaxHighlight::NoColor => "none",
        _ => panic!("unexpected fixture highlight: {highlight:?}"),
    }
}

fn varnode_id(identity: i64) -> &'static str {
    match identity {
        -1 => "none",
        0 => "obj0",
        _ => panic!("unregistered fixture Varnode identity: {identity}"),
    }
}

fn op_id(identity: i64) -> &'static str {
    match identity {
        -1 => "none",
        1 => "obj1",
        _ => panic!("unregistered fixture PcodeOp identity: {identity}"),
    }
}

impl RecordingEmit {
    fn append(
        &mut self,
        kind: &str,
        text: &str,
        highlight: SyntaxHighlight,
        varnode_identity: i64,
        op_identity: i64,
        group: &str,
    ) {
        self.events.push(format!(
            "event={}|kind={kind}|text_hex={}|highlight={}|vn={}|op={}|group={group}",
            self.events.len(),
            hex_text(text),
            color_name(highlight),
            varnode_id(varnode_identity),
            op_id(op_identity),
        ));
    }
}

impl Emit for RecordingEmit {
    fn print(&mut self, text: &str) {
        self.append("syntax", text, SyntaxHighlight::NoColor, -1, -1, "none");
    }

    fn open_group(&mut self) -> i32 {
        let id = self.next_group;
        self.next_group += 1;
        self.append(
            "open_group",
            "",
            SyntaxHighlight::NoColor,
            -1,
            -1,
            &format!("g{id}"),
        );
        id
    }

    fn close_group(&mut self, id: i32) {
        assert!(
            (0..self.next_group).contains(&id),
            "unknown fixture group id"
        );
        self.append(
            "close_group",
            "",
            SyntaxHighlight::NoColor,
            -1,
            -1,
            &format!("g{id}"),
        );
    }

    fn begin_block(&mut self) {}
    fn end_block(&mut self) {}
    fn begin_function(&mut self) {}
    fn end_function(&mut self) {}
    fn tag_type(&mut self, _text: &str, _id: u64) {}
    fn tag_variable(&mut self, _text: &str, _id: u64) {
        panic!("metadata-aware variable bridge was bypassed");
    }

    fn tag_variable_with_metadata(
        &mut self,
        text: &str,
        highlight: SyntaxHighlight,
        varnode_identity: i64,
        op_identity: i64,
    ) {
        self.append(
            "variable",
            text,
            highlight,
            varnode_identity,
            op_identity,
            "none",
        );
    }

    fn tag_op(&mut self, _text: &str) {}
    fn tag_field(&mut self, _text: &str, _id: u64) {}
    fn tag_func_name(&mut self, _text: &str, _id: u64) {}
    fn tag_comment(&mut self, _text: &str) {}
    fn tag_label(&mut self, _text: &str) {}
    fn tag_case_label(&mut self, _text: &str) {}
    fn into_any(self: Box<Self>) -> Box<dyn std::any::Any> {
        self
    }
}

fn push_hidden_atom(
    revpol: &mut Vec<ReversePolish>,
    nodepend: &mut Vec<NodePending>,
    pending: &mut usize,
    tokens: &[OpToken],
    emit: &mut RecordingEmit,
    atom: &Atom,
    consuming_op_identity: i64,
) {
    rpn_push_op(
        revpol,
        nodepend,
        pending,
        tokens,
        emit,
        0,
        consuming_op_identity,
    );
    rpn_push_atom(revpol, nodepend, pending, tokens, emit, atom);
}

fn main() {
    // A single global creation-order namespace distinguishes the two metadata
    // domains and catches accidental vn/op field interchange.
    const VARNODE_CREATION_ID: i64 = 0;
    const CONSUMING_OP_CREATION_ID: i64 = 1;
    struct IdentityGraph {
        varnode: i64,
        consuming_op: i64,
        consuming_op_input0: i64,
    }
    let graph = IdentityGraph {
        varnode: VARNODE_CREATION_ID,
        consuming_op: CONSUMING_OP_CREATION_ID,
        consuming_op_input0: VARNODE_CREATION_ID,
    };
    assert_ne!(graph.varnode, graph.consuming_op);
    assert_eq!(graph.consuming_op_input0, graph.varnode);

    // Exact counterpart of Ghidra PrintC::hidden at printc.cc:23.  Construct
    // the fixture input explicitly so this projection is independent of
    // convenience-constructor coverage.
    let tokens = vec![OpToken {
        print1: String::new(),
        print2: String::new(),
        stage: 1,
        precedence: 70,
        associative: false,
        type_: TokenType::HiddenFunction,
        spacing: 0,
        bump: 0,
        negate: -1,
    }];
    let atom = Atom::with_op_vn_int(
        "'\\0'",
        TagType::VarToken,
        SyntaxHighlight::ConstColor,
        CONSUMING_OP_CREATION_ID,
        VARNODE_CREATION_ID,
        0,
    );
    let syntax = Atom::new(";", TagType::Syntax, SyntaxHighlight::NoColor);

    let mut revpol = Vec::new();
    let mut nodepend = Vec::new();
    let mut pending = 0;
    let mut emit = RecordingEmit::default();

    push_hidden_atom(
        &mut revpol,
        &mut nodepend,
        &mut pending,
        &tokens,
        &mut emit,
        &atom,
        graph.consuming_op,
    );
    push_hidden_atom(
        &mut revpol,
        &mut nodepend,
        &mut pending,
        &tokens,
        &mut emit,
        &atom,
        graph.consuming_op,
    );
    rpn_push_atom(
        &mut revpol,
        &mut nodepend,
        &mut pending,
        &tokens,
        &mut emit,
        &syntax,
    );

    assert!(
        revpol.is_empty(),
        "fixture must completely drain the RPN stack"
    );
    assert!(nodepend.is_empty(), "fixture does not use pending Varnodes");
    assert_eq!(pending, 0);

    println!(
        "schema=1|fixture=PRINTLANGUAGE-ATOM-METADATA-0001|\
         oracle=e40ed13014025f82488b1f8f7bca566894ac376b|\
         covered_projection=MATCH|overall=MISMATCH"
    );
    for event in emit.events {
        println!("{event}");
    }
}
